### ISSUE TITLE
Jupyter does not recognize local ip under some circumstances

### ISSUE THREAD

**Issue by asteppke** [jupyter/jupyter_client#466]
I am running into this odd issue that jupyter refuses to run because it does not recognize the localhost anymore under special circumstances.

The error message I receive is
```
 File "c:\users\alexander\anaconda3\lib\site-packages\jupyter_client\manager.py", line 236, in start_kernel
		"Currently valid addresses are: %s" % (self.ip, local_ips())
	RuntimeError: Can only launch a kernel on a local interface. 
This one is not: 127.0.0.1.Make sure that the '*_address' attributes are configured properly. 
Currently valid addresses are: ['10.1x.xxx.xxx', '0.0.0.0', '']
```

This occurs when starting the notebook from a different process (treon, https://github.com/ReviewNB/treon, which calls nbconvert internally to run the notebook), not when starting the notebook from the command line. When the external network interface is switched off the issue disappears. 

I am reporting this here because in this case treon is the calling program but a similar issue was previously reported in the Spyder issue tracker so it does not seem to be specific to the caller (https://github.com/winpython/winpython/issues/582 when Spyder is starting the jupyter client, closed but unresolved).

I am not sure why the jupyter_client manager does not accept 127.0.0.1 as a local interface and why this is only an issue when another external ip exists on the same machine. Furthermore to make this even more of an Heisenbug than it already is: to trigger this behavior two jupyter instances must be called at the same time, i.e. when treon is restricted to only one thread the jupyter client manager does not complain anymore.

This all occurs on a Windows 10 machine, with the current Anaconda distribution using jupyter_client 5.3.1.

**Comment by asteppke** [comment]
After poking around a bit the issue seems to be connected to `localinterfaces.py`. When using its public interface, that is `is_local_ip(addrs)` the list of local ip addresses gets populated and should include `127.0.0.1`. 

Internally this is assured by setting a global variable and appending to the list and the whole procedure should only run once (this is assured via a decorator). Now my guess is that in the case some concurrent calls this fails, the global variable is set but the internal functions execute twice, not once and therefore the list of local ips is not correct from the second call on.

**Comment by amit1rrr** [comment]
I'm the maintainer of [treon](https://github.com/reviewNB/treon) which uses jupyter_client. As @asteppke mentioned, there's something peculiar happening while running jupyter_client in multiple parallel threads on windows. It's not picking up `127.0.0.1` as local ip. 

**Possible fix** - what if we always include `127.0.0.1` on the line below then probably don't even need the `if` clause above it. (https://github.com/jupyter/jupyter_client/blob/2dc0e4e826b7547d82d7d17594fd955adffa2dd3/jupyter_client/localinterfaces.py#L86) 

I don't have complete understanding of how LOCALHOST plays into `localinterfaces.py` so forgive me if this is a bad suggestion. If there's anything that clients should do while using jupyter_client then please let us know.

@minrk Any thoughts here?

**Comment by MSeal** [comment]
Sorry we missed this thread before.

I'm kind of surprised it doesn't always include `127.0.0.1` to be honest, though this code hasn't been touched in 5 years so it predates my involvement. I think changing `if not LOCALHOST:` to `if '127.0.0.1' not in LOCALHOST:` would be reasonable. Do you want to try making this change locally and seeing if it resolves the reported issue for you?

**Comment by MSeal** [comment]
If I'm not responsive after this (going to be unavailable for a while soon) and other maintainers don't chime in try pinging @ willingc and she can help find available devs

**Comment by amit1rrr** [comment]
Thanks @MSeal 

@asteppke I don't have a windows machine to reproduce the issue. I know it's been a while but is it possible for you to try out the following suggestion? I will lay out the steps below if that makes it easy.

- Clone jupyter_client repository
- Change the following line to `if '127.0.0.1' not in LOCALHOST:` as suggested by @MSeal 
https://github.com/jupyter/jupyter_client/blob/2dc0e4e826b7547d82d7d17594fd955adffa2dd3/jupyter_client/localinterfaces.py#L82
- Install your locally edited jupyter_client with `pip install -e .`
- Try running treon in multi-threaded mode

**Comment by GenevieveBuckley** [comment]
It seems the fix isn't quite `if '127.0.0.1' not in LOCALHOST:`, but a variation on that suggestion  `if not LOCALHOST or LOCALHOST == '127.0.0.1':` has worked well on my machine.
